spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerOperations.java

```diff
@@ -78,29 +79,27 @@
 	<T> List<T> read(Class<T> entityClass, KeySet keys);
 
 	/**
-	 * Finds objects by using an SQL statement.
-	 * @param entityClass the type of object to retrieve.
-	 * @param sql the SQL string to execute. this string can have Cloud Spanner param tags.
-	 * @param tags the names of the tags to use
-	 * @param params the values to attach those tags, in the same order.
-	 * @param options Cloud Spanner read options with which to conduct the read operation.
-	 * @param <T> the type of object to retrieve.
-	 * @return a list of the objects found. If no keys could be found the list will be
-	 * empty.
+	 * Executes a given query string with tags and parameters and applies a given function
+	 * to each row of the result.
+	 * @param rowFunc the function to apply to each row of the result.
+	 * @param statement the SQL statement used to select the objects.
+	 * @param options the options with which to run this query.
+	 * @return a list of the rows each transformed with the given function.
 	 */
-	<T> List<T> query(Class<T> entityClass, String sql, List<String> tags,
-			Object[] params,
+	<A> List<A> query(Function<Struct, A> rowFunc, Statement statement,
```

Comment:
Have a look at `JdbcTemplate` (or `CqlTemplate` vs. `CassandraTemplate`) as these type of functions are typically applied on a lower abstraction level (`Statement` level vs. logical operations).